File: 2_medium/0001_1000/00053_maximum_subarray/solution.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {

        int sum{0}, max{INT_MIN}, nums_size{static_cast<int>(nums.size())};

        for(int i{0}; i != nums_size; ++i) {
            sum += nums[i];
            max = std::max(max, sum);
            
            // If sum be less than 0, then reset sum to 0 and start again
            sum = std::max(sum, 0);
        }

        return max;
    }
};

File: 2_medium/0001_1000/00053_maximum_subarray/solution.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {

        let mut sum = 0_i32;
        let mut max = i32::MIN;
        let nums_len = nums.len();

        for i in 0..nums_len {
            sum += nums[i];
            max = max.max(sum);

            // If sum be less than 0, then reset sum to 0 and start again
            sum = sum.max(0);
        }

        max
    }
}
